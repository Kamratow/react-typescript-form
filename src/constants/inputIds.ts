export default class inputIds {
    static 'EMAIL' = 'emailInput';
    static 'PASSWORD' = 'passwordInput';
    static 'FULL_NAME' = 'fullNameInput';
    static 'CONFIRM_PASSWORD' = 'confirmPasswordInput';
}