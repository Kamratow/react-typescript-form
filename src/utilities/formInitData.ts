import { IFormInput } from "../models/formInput.model";
import inputIds from "../constants/inputIds";
import textConstants from "../constants/textConstants";

export const getInitialSignInFormData = (): IFormInput[] => {
    return [
        {
            id: inputIds.EMAIL,
            value: "",
            type: "email",
            label: textConstants.EMAIL_INPUT_LABEL,
            errorMessage: "",
        },
        {
            id: inputIds.PASSWORD,
            value: "",
            type: "password",
            label: textConstants.PASSWORD_INPUT_LABEL,
            errorMessage: "",
        },
    ];
};

export const getInitialSignUpFormData = (): IFormInput[] => {
    return [
        {
            id: inputIds.EMAIL,
            value: "",
            type: "email",
            label: textConstants.EMAIL_INPUT_LABEL,
            errorMessage: "",
        },
        {
            id: inputIds.FULL_NAME,
            value: "",
            type: "text",
            label: textConstants.FULL_NAME_INPUT_LABEL,
            errorMessage: "",
        },
        {
            id: inputIds.PASSWORD,
            value: "",
            type: "password",
            label: textConstants.PASSWORD_INPUT_LABEL,
            errorMessage: "",
        },
        {
            id: inputIds.CONFIRM_PASSWORD,
            value: "",
            type: "password",
            label: textConstants.CONFIRM_PASSWORD_INPUT_LABEL,
            errorMessage: "",
        }
    ];
};