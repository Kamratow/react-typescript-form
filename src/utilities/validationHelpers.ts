import {IFormInput} from "../models/formInput.model";
import inputIds from "../constants/inputIds";

export const validateAndParseInputData = (newValue: string, formInput: IFormInput) => {
    let parsedAndValidatedInput: IFormInput = {
        ...formInput,
        value: newValue,
        errorMessage: ''
    };

    //validate if input is empty
    if (newValue === '') {
        if (parsedAndValidatedInput.id === inputIds.CONFIRM_PASSWORD) {
            parsedAndValidatedInput.errorMessage = "Please confirm your password";
        } else {
            parsedAndValidatedInput.errorMessage = `${parsedAndValidatedInput.label} is required`;
        }

        return parsedAndValidatedInput;
    }

    //validate if value matches input type
    switch (formInput.type) {
        case 'email':
            parsedAndValidatedInput = validateEmailValue(parsedAndValidatedInput);
            break;
        case 'password':
            parsedAndValidatedInput = validatePasswordValue(parsedAndValidatedInput);
            break;
        default:
            break;
    }

    //check if input is full name
    if (parsedAndValidatedInput.id === inputIds.FULL_NAME) {
        parsedAndValidatedInput = validateFullNameValue(parsedAndValidatedInput);
    }

    return parsedAndValidatedInput;
};

export const validateEmailValue = (formInput: IFormInput): IFormInput => {
    const emailRegex = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

    if (!emailRegex.test(formInput.value)) {
        formInput.errorMessage = 'Invalid e-mail address';
    }

    return formInput;
};

export const validatePasswordValue = (formInput: IFormInput): IFormInput => {
    if (formInput.value.length < 8) {
        formInput.errorMessage = 'Password is too short';
    }

    return formInput;
};

export const validateFullNameValue = (formInput: IFormInput): IFormInput => {
    const fullNameRegex = /([a-zA-Z]+\s+[a-zA-Z]+)+/;

    if (!fullNameRegex.test(formInput.value)) {
        formInput.errorMessage = 'Missing last name';
    }

    return formInput;
};

export const checkIsFormDataValid = (formData: IFormInput[]): boolean => {
    return formData.filter(singleInput => singleInput.value !== '').length === formData.length && !formData.find(singleInput => singleInput.errorMessage !== '');
};

export const validateConfirmPasswordValue = (formData: IFormInput[]): IFormInput[] => {
    const validatedFormData = [...formData];
    const passwordInput = validatedFormData.find(singleInput => singleInput.id === inputIds.PASSWORD);
    const confirmPasswordInput = validatedFormData.find(singleInput => singleInput.id === inputIds.CONFIRM_PASSWORD);

    if (passwordInput && confirmPasswordInput && passwordInput.value !== '' && confirmPasswordInput.value !== '' && passwordInput.value !== confirmPasswordInput.value) {
        validatedFormData[validatedFormData.findIndex(singleInput => singleInput.id === inputIds.CONFIRM_PASSWORD)].errorMessage = 'Please confirm your password';
    }

    return validatedFormData;
};
