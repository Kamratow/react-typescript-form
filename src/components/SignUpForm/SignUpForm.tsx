import React, { useState } from "react";

import Header from "../common/Header/Header";
import Form from "../common/Form/Form";
import StandardField from "../common/StandardField/StandardField";
import FormLink from "../common/FormLink/FormLink";
import SubmitButton from "../common/SubmitButton/SubmitButton";

import { getInitialSignUpFormData } from "../../utilities/formInitData";
import {
    validateAndParseInputData,
    checkIsFormDataValid,
    validateConfirmPasswordValue,
} from "../../utilities/validationHelpers";

import textConstants from "../../constants/textConstants";

import { IFormInput } from "../../models/formInput.model";

const SignUpForm: React.FunctionComponent = () => {
    const [formData, setFormData] = useState<IFormInput[]>(
        getInitialSignUpFormData()
    );

    const handleInputChange = (
        newValue: string,
        formInputData: IFormInput,
        inputIndex: number
    ): void => {
        let newFormData = [...formData];
        const newFormInputData = validateAndParseInputData(
            newValue,
            formInputData
        );

        newFormData[inputIndex] = {
            ...newFormInputData,
        };

        if (formInputData.type === "password") {
            newFormData = validateConfirmPasswordValue(newFormData);
        }

        setFormData(newFormData);
    };

    const getFormInputsToRender = () => {
        return formData.map((singleInput, index) => {
            return (
                <StandardField
                    key={singleInput.id}
                    id={singleInput.id}
                    label={singleInput.label}
                    type={singleInput.type}
                    value={singleInput.value}
                    changeHandler={(newInputValue: string) =>
                        handleInputChange(newInputValue, singleInput, index)
                    }
                    errorMessage={singleInput.errorMessage}
                />
            );
        });
    };

    return (
        <>
            <Header headerContent={textConstants.SIGNUP_HEADER} />
            <Form>
                {getFormInputsToRender()}
                <SubmitButton
                    disabled={!checkIsFormDataValid(formData)}
                    buttonLabel={textConstants.FORM_SUBMIT_BUTTON}
                />
            </Form>
            <FormLink
                description={textConstants.SIGNUP_FORMLINK_DESCRIPTION}
                linkPath={textConstants.SIGNUP_FORMLINK_PATH}
                linkText={textConstants.SIGNUP_FORMLINK_LINK_CONTENT}
            />
        </>
    );
};

export default SignUpForm;
