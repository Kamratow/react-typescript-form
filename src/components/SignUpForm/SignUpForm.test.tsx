import React from "react";
import { shallow } from "enzyme";

import SignUpForm from "./SignUpForm";

import * as formInitData from "../../utilities/formInitData";
import inputIds from "../../constants/inputIds";

import {
    validSignUpFormData,
    invalidSignUpFormData,
} from "./testMocks/signUpformDataMock";

describe("SignUpForm", () => {
    it("should render with proper input configuration", () => {
        const wrapper = shallow(<SignUpForm />);

        expect(wrapper).toMatchSnapshot();
    });

    describe("FullNameInput", () => {
        let wrapper: any;
        let fullNameInput: any;

        const getFullNameInput = (inputWrapper: any) => {
            return inputWrapper
                .findWhere(
                    (component: any) =>
                        component.name() === "StandardField" &&
                        component.prop("id") === inputIds.FULL_NAME
                )
                .at(0);
        };

        beforeEach(() => {
            wrapper = shallow(<SignUpForm />);
            fullNameInput = getFullNameInput(wrapper);
        });

        it("should be properly validated when full name is incorrect", () => {
            const expectedErrorMessage = "Missing last name";
            const invalidFullName = "test";

            fullNameInput.prop("changeHandler")(invalidFullName);

            fullNameInput = getFullNameInput(wrapper);

            expect(fullNameInput.prop("value")).toBe(invalidFullName);
            expect(fullNameInput.prop("errorMessage")).toBe(
                expectedErrorMessage
            );
        });

        it("should be properly validated when full name is correct", () => {
            const validFullName = "test name";

            fullNameInput.prop("changeHandler")(validFullName);

            fullNameInput = getFullNameInput(wrapper);

            expect(fullNameInput.prop("value")).toBe(validFullName);
            expect(fullNameInput.prop("errorMessage")).toBe("");
        });

        it("should be properly validated when full name is empty", () => {
            const expectedErrorMessage = "Full name is required";
            const invalidFullName = "";

            fullNameInput.prop("changeHandler")(invalidFullName);

            fullNameInput = getFullNameInput(wrapper);

            expect(fullNameInput.prop("value")).toBe(invalidFullName);
            expect(fullNameInput.prop("errorMessage")).toBe(
                expectedErrorMessage
            );
        });
    });

    describe("ConfirmPassword", () => {
        const getConfirmPasswordInput = (inputWrapper: any) => {
            return inputWrapper
                .findWhere(
                    (component: any) =>
                        component.name() === "StandardField" &&
                        component.prop("id") === inputIds.CONFIRM_PASSWORD
                )
                .at(0);
        };

        beforeEach(() => {
            jest.clearAllMocks();
        });

        it("should be properly validate when confirm password is empty", () => {
            const expectedErrorMessage = "Please confirm your password";
            const invalidConfirmPassword = "";
            const wrapper = shallow(<SignUpForm />);
            let confirmPasswordInput = getConfirmPasswordInput(wrapper);

            confirmPasswordInput.prop("changeHandler")(invalidConfirmPassword);

            confirmPasswordInput = getConfirmPasswordInput(wrapper);

            expect(confirmPasswordInput.prop("value")).toBe(
                invalidConfirmPassword
            );
            expect(confirmPasswordInput.prop("errorMessage")).toBe(
                expectedErrorMessage
            );
        });

        it("should be properly validated when confirm password is different than password", () => {
            jest.spyOn(
                formInitData,
                "getInitialSignUpFormData"
            ).mockImplementation(() => invalidSignUpFormData);

            const expectedErrorMessage = "Please confirm your password";
            const invalidConfirmPassword = "1234";
            const wrapper = shallow(<SignUpForm />);
            let confirmPasswordInput = getConfirmPasswordInput(wrapper);

            confirmPasswordInput.prop("changeHandler")(invalidConfirmPassword);

            confirmPasswordInput = getConfirmPasswordInput(wrapper);

            expect(confirmPasswordInput.prop("value")).toBe(
                invalidConfirmPassword
            );
            expect(confirmPasswordInput.prop("errorMessage")).toBe(
                expectedErrorMessage
            );
        });

        it("should be properly validated when confirm password is equal to password", () => {
            jest.spyOn(
                formInitData,
                "getInitialSignUpFormData"
            ).mockImplementation(() => invalidSignUpFormData);

            const expectedErrorMessage = "";
            const validConfirmPassword = "12345678";
            const wrapper = shallow(<SignUpForm />);
            let confirmPasswordInput = getConfirmPasswordInput(wrapper);

            confirmPasswordInput.prop("changeHandler")(validConfirmPassword);

            confirmPasswordInput = getConfirmPasswordInput(wrapper);

            expect(confirmPasswordInput.prop("value")).toBe(
                validConfirmPassword
            );
            expect(confirmPasswordInput.prop("errorMessage")).toBe(
                expectedErrorMessage
            );
        });
    });

    describe("SubmitButton", () => {
        it("should be disabled if form data is invalid", () => {
            jest.spyOn(
                formInitData,
                "getInitialSignUpFormData"
            ).mockImplementation(() => invalidSignUpFormData);

            const wrapper = shallow(<SignUpForm />);

            expect(wrapper.find("SubmitButton").prop("disabled")).toBe(true);
        });

        it("should not be disabled if form data is valid", () => {
            jest.spyOn(
                formInitData,
                "getInitialSignUpFormData"
            ).mockImplementation(() => validSignUpFormData);

            const wrapper = shallow(<SignUpForm />);

            expect(wrapper.find("SubmitButton").prop("disabled")).toBe(false);
        });
    });
});
