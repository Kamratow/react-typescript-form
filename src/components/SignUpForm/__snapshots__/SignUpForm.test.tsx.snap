// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpForm should render with proper input configuration 1`] = `
<Fragment>
  <Header
    headerContent="Sign up"
  />
  <Form>
    <StandardField
      changeHandler={[Function]}
      errorMessage=""
      id="emailInput"
      key="emailInput"
      label="E-mail address"
      type="email"
      value=""
    />
    <StandardField
      changeHandler={[Function]}
      errorMessage=""
      id="fullNameInput"
      key="fullNameInput"
      label="Full name"
      type="text"
      value=""
    />
    <StandardField
      changeHandler={[Function]}
      errorMessage=""
      id="passwordInput"
      key="passwordInput"
      label="Password"
      type="password"
      value=""
    />
    <StandardField
      changeHandler={[Function]}
      errorMessage=""
      id="confirmPasswordInput"
      key="confirmPasswordInput"
      label="Confirm your password"
      type="password"
      value=""
    />
    <SubmitButton
      buttonLabel="Continue"
      disabled={true}
    />
  </Form>
  <FormLink
    description="Already using our app?"
    linkPath="/signin"
    linkText="Click here to sign in"
  />
</Fragment>
`;
